dpnp/dpnp_iface_linearalgebra.py

```diff
@@ -269,63 +330,325 @@ def matmul(x1, x2, out=None, **kwargs):
+            result = dpnp.astype(result, res_dtype)
+
+        if out is not None:
+            return dpnp.get_result_array(result, out, casting=casting)
+
+        # If `order` was not passed as default
+        # we must copy `result` to match the passed `order`.
+        if order not in ["k", "K"]:
+            return dpnp.copy(result, order=order)
+
+        return result
+
+
+def dpnp_matmul_batch(
+    exec_q, x1, x2, res, x1_is_2D, x2_is_2D, ht_copy_ev_x1, ht_copy_ev_x2
+):
+    # If input array is F-contiguous, we need to change the order to C-contiguous.
+    # because mkl::gemm_bacth needs each 2D array to be F-contiguous but
+    # when the input array is F-contiguous, the data of 2D array
+    # that needs to be called in mkl::gemm_batch are not contiguous.
+    copy_ev_x1 = dpctl.SyclEvent()
+    if not x1.flags.c_contiguous:
+        x1_copy = dpnp.empty_like(x1, order="C")
+        (
+            ht_copy_ev_x1,
+            copy_ev_x1,
+        ) = ti._copy_usm_ndarray_into_usm_ndarray(
+            src=dpnp.get_usm_ndarray(x1),
+            dst=x1_copy.get_array(),
+            sycl_queue=x1.sycl_queue,
+        )
+        x1 = x1_copy
+
+    copy_ev_x2 = dpctl.SyclEvent()
+    if not x2.flags.c_contiguous:
+        x2_copy = dpnp.empty_like(x2, order="C")
+        (
+            ht_copy_ev_x2,
+            copy_ev_x2,
+        ) = ti._copy_usm_ndarray_into_usm_ndarray(
+            src=dpnp.get_usm_ndarray(x2),
+            dst=x2_copy.get_array(),
+            sycl_queue=x2.sycl_queue,
+        )
+        x2 = x2_copy
+
+    x1_strides = x1.strides
+    x2_strides = x2.strides
+    res_strides = res.strides
+
+    # When shape along any particular dimension is 1,
+    # the stride along that dimension is not a
+    # meaningful number and is undefined. Here, we
+    # standardizing strides before continuing,
+    # setting stride to 0 if the shape along that axis is <=1
+    if x1_is_2D:
+        x1_strides = tuple(
+            str_i if sh_i > 1 else 0
+            for sh_i, str_i in zip(x1.shape, x1_strides)
+        )
+    if x2_is_2D:
+        x2_strides = tuple(
+            str_i if sh_i > 1 else 0
+            for sh_i, str_i in zip(x2.shape, x2_strides)
+        )
+
+    batch_size = res.shape[:-2][0]
+    m = x1.shape[-2]
+    n = x2.shape[-1]
+    k = x1.shape[-1]
+
+    stridea = x1_strides[0]
+    strideb = x2_strides[0]
+    stridec = res_strides[-3]
+
+    if x1.ndim > 3:
+        iter = ti._contract_iter2(
+            res.shape[:-2], x1_strides[:-2], x2_strides[:-2]
+        )
 
-    return call_origin(numpy.matmul, x1, x2, out=out, **kwargs)
+        if len(iter[0]) != 1:
+            raise ValueError("Input arrays cannot be used in gemm_batch")
+        batch_size = iter[0][0]
+        stridea = iter[1][0]
+        strideb = iter[3][0]
+
+    ht_blas_ev, _ = bi._gemm_batch(
+        exec_q,
+        dpnp.get_usm_ndarray(x1),
+        dpnp.get_usm_ndarray(x2),
+        dpnp.get_usm_ndarray(res),
+        m,
```

Comment:
Do we need to have `m, n, k` and `lda, ldb, ldc` as input arguments? Looks we can take it from input arrays already passed to `_gemm_batch`.